src/main/scala/mesosphere/marathon/api/akkahttp/PathMatchers.scala

```diff
@@ -0,0 +1,74 @@
+package mesosphere.marathon
+package api.akkahttp
+
+import akka.http.scaladsl.model.Uri.Path
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.state.{ Group, PathId, RootGroup, Timestamp }
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.PathMatcher.{ Matched, Unmatched }
+import akka.http.scaladsl.server.PathMatcher1
+
+import scala.annotation.tailrec
+
+object PathMatchers {
+
+  /**
+    * Matches the remaining path and transforms it into task id
+    */
+  final val RemainingTaskId = Remaining.map(s => Task.Id(s))
+
+  /**
+    * Tries to match the remaining path as Timestamp
+    */
+  final val Version = Segment.flatMap(string =>
+    try Some(Timestamp(string))
+    catch { case _: IllegalArgumentException ⇒ None }
+  )
+
+  /**
+    * Given the current root group, only match and consume an existing appId
+    *
+    * This is useful because our v2 API has an unfortunate design decision which leads to ambiguity in our URLs, such as:
+    *
+    *   POST /v2/apps/my-group/restart/restart
+    *
+    * The intention here is to restart the app named "my-group/restart"
+    *
+    * Given the url above, this matcher will only consume "my-group/restart" from the path,
+    * leaving the rest of the matcher to match the rest
+    */
+  case class ExistingAppPathId(rootGroup: RootGroup) extends PathMatcher1[PathId] {
```

Comment:
How does it related to `extractExistingAppId`?

Reply:
This is a `PathMatcher`. It can match a correct appId. `extractExistingAppId` is a directive which uses it to extract appId or to reject with our own rejection in case appId is not found.